### PR TITLE
[Core Plugin] adopt more change from pr 3414

### PR DESCRIPTION
**Pull request by Jim8y**
# Description

3rd part of Pr #3414 and credits belong to @cschuchardt88 

- Added IEnumerable<KeyValuePair<byte[], byte[]>> to Store.cs to Iterate over the whole store.
- Changed namespace from Neo.IO.Data.LevelDB to Neo.IO.Storage.LevelDB
- Update IntPtr to nint 

Fixes # #3414 

## Type of change

- [ ] Optimization (the change is only an optimization)
- [x] Style (the change is only a code style for better maintenance or standard purpose)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/Plugins/LevelDBStore/IO/Data/LevelDB/DB.cs [hunk 1]
```diff
@@ -122,5 +124,22 @@ public void Write(WriteOptions options, WriteBatch write_batch)
             Native.leveldb_write(Handle, options.Handle, write_batch.Handle, out var error);
             NativeHelper.CheckError(error);
         }
+
+        public IEnumerator<KeyValuePair<byte[], byte[]>> GetEnumerator()
+        {
+            using var sn = GetSnapshot();
+            using var iterator = NewIterator(new ReadOptions { Snapshot = sn });
+            iterator.SeekToFirst();
```
Comment (shargon) [inline src/Plugins/LevelDBStore/IO/Data/LevelDB/DB.cs:132]: This call is required?

Comment (Jim8y) [reply src/Plugins/LevelDBStore/IO/Data/LevelDB/DB.cs:132]: @cschuchardt88

File: src/Plugins/LevelDBStore/Plugins/Storage/Store.cs [hunk 2]
```diff
@@ -53,6 +54,12 @@ public bool TryGet(byte[] key, out byte[] value)
...
 
         public IEnumerable<(byte[], byte[])> Seek(byte[] prefix, SeekDirection direction = SeekDirection.Forward) =>
-            _db.Seek(ReadOptions.Default, prefix, direction, (k, v) => (k, v));
+            _db.Seek(ReadOptions.Default, prefix, direction);
+
+        public IEnumerator<KeyValuePair<byte[], byte[]>> GetEnumerator() =>
+            _db.GetEnumerator();
+
+        IEnumerator IEnumerable.GetEnumerator() =>
+            GetEnumerator();
```
[the remark this answers is not in the document]
Comment (cschuchardt88) [reply src/Plugins/LevelDBStore/Plugins/Storage/Store.cs]: these are two different methods. `GetEnumerator();` is correct

File: src/Plugins/LevelDBStore/IO/Data/LevelDB/DB.cs [hunk 3]
```diff
@@ -122,5 +124,22 @@ public void Write(WriteOptions options, WriteBatch write_batch)
...
+            while (iterator.Valid())
+            {
+                yield return new KeyValuePair<byte[], byte[]>(iterator.Key(), iterator.Value());
```
[the remark this answers is not in the document]
Comment (shargon) [reply src/Plugins/LevelDBStore/IO/Data/LevelDB/DB.cs:135]: I prefer to have the snapshot in an argument, maybe we don't wan't to use it, and avoid ` IEnumerable<KeyValuePair<byte[], byte[]>>`

File: src/Plugins/LevelDBStore/Plugins/Storage/Store.cs [hunk 4]
```diff
@@ -53,6 +54,12 @@ public bool TryGet(byte[] key, out byte[] value)
...
 
         public IEnumerable<(byte[], byte[])> Seek(byte[] prefix, SeekDirection direction = SeekDirection.Forward) =>
-            _db.Seek(ReadOptions.Default, prefix, direction, (k, v) => (k, v));
+            _db.Seek(ReadOptions.Default, prefix, direction);
+
+        public IEnumerator<KeyValuePair<byte[], byte[]>> GetEnumerator() =>
+            _db.GetEnumerator();
+
+        IEnumerator IEnumerable.GetEnumerator() =>
+            _db.GetEnumerator();
```
Comment (shargon) [inline src/Plugins/LevelDBStore/Plugins/Storage/Store.cs:63]: Same here

File: src/Plugins/LevelDBStore/IO/Data/LevelDB/DB.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (cschuchardt88) [reply src/Plugins/LevelDBStore/IO/Data/LevelDB/DB.cs:135]: It's helpful for debugging. Better than being in the dark.

[the remark this answers is not in the document]
Comment (Jim8y) [reply src/Plugins/LevelDBStore/IO/Data/LevelDB/DB.cs:135]: > It's helpful for debugging. Better than being in the dark.

Will this cause state difference or behavior difference?

[the remark this answers is not in the document]
Comment (shargon) [reply src/Plugins/LevelDBStore/IO/Data/LevelDB/DB.cs:135]: We should receive the snapshot  or null or we can have a big mistakes

[the remark this answers is not in the document]
Comment (shargon) [reply src/Plugins/LevelDBStore/IO/Data/LevelDB/DB.cs:135]: But imagine that we use this method in a loop, when we change the data, we would not notice this error

[the remark this answers is not in the document]
Comment (cschuchardt88) [reply src/Plugins/LevelDBStore/IO/Data/LevelDB/DB.cs:135]: than don't use `snapshot` just use 
```csharp
using var iterator = NewIterator(ReadOptions.Default);
```

**Review by cschuchardt88 — APPROVED** [review]

File: src/Plugins/LevelDBStore/IO/Data/LevelDB/DB.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (shargon) [reply src/Plugins/LevelDBStore/IO/Data/LevelDB/DB.cs:135]: Easy to forget

File: src/Plugins/LevelDBStore/Plugins/Storage/Snapshot.cs [hunk 5]
```diff
@@ -44,11 +44,12 @@ public void Delete(byte[] key)
         public void Dispose()
         {
             _snapshot.Dispose();
+            _options.Dispose();
```
Comment (shargon) [inline src/Plugins/LevelDBStore/Plugins/Storage/Snapshot.cs:47]: This memory leak is present in master

**Review by shargon — APPROVED** [review]

File: src/Plugins/LevelDBStore/IO/Data/LevelDB/DB.cs [hunk 6]
```diff
@@ -10,15 +10,17 @@
...
 
-namespace Neo.IO.Data.LevelDB
+namespace Neo.IO.Storage.LevelDB
 {
     /// <summary>
     /// A DB is a persistent ordered map from keys to values.
     /// A DB is safe for concurrent access from multiple threads without any external synchronization.
     /// </summary>
-    public class DB : LevelDBHandle
+    public class DB : LevelDBHandle, IEnumerable<KeyValuePair<byte[], byte[]>>
```
[the remark this answers is not in the document]
Comment (shargon) [reply src/Plugins/LevelDBStore/IO/Data/LevelDB/DB.cs]: This can come in different PR

File: src/Plugins/LevelDBStore/Plugins/Storage/Store.cs [hunk 7]
```diff
@@ -18,18 +18,24 @@ namespace Neo.Plugins.Storage
...
         public void Delete(byte[] key)
         {
             _db.Delete(WriteOptions.Default, key);
         }
 
-        public void Dispose() => _db.Dispose();
+        public void Dispose()
+        {
+            _db.Dispose();
+            _options.Dispose();
```
Comment (shargon) [inline src/Plugins/LevelDBStore/Plugins/Storage/Store.cs:37]: This one also, but one per node

File: src/Plugins/LevelDBStore/IO/Data/LevelDB/DB.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (shargon) [reply src/Plugins/LevelDBStore/IO/Data/LevelDB/DB.cs:138]: We can discuss in a different PR if we create the `IEnumerable<KeyValuePair<byte[], byte[]>>` , this one is ready to merge

File: src/Plugins/LevelDBStore/IO/Data/LevelDB/DB.cs [hunk 8]
```diff
@@ -122,5 +125,19 @@ public void Write(WriteOptions options, WriteBatch write_batch)
...
         }
+
+        public IEnumerable<KeyValuePair<byte[], byte[]>> GetAll(Snapshot? snapshot = null)
+        {
+            using var options = new ReadOptions();
+            if (snapshot != null) options.Snapshot = snapshot;
+
+            using var iterator = NewIterator(options);
+            iterator.SeekToFirst();
+            while (iterator.Valid())
```
Comment (nan01ab) [inline src/Plugins/LevelDBStore/IO/Data/LevelDB/DB.cs:136]: Why is this `while`, and below is `for`. Might it be better to keep a consistent style?

Comment (cschuchardt88) [reply src/Plugins/LevelDBStore/IO/Data/LevelDB/DB.cs:136]: He copied mine. but `for` would be better.

Comment (Jim8y) [reply src/Plugins/LevelDBStore/IO/Data/LevelDB/DB.cs:136]: will be fixed in another pr.

**Review by cschuchardt88 — APPROVED** [review]

**Pull request merged**